### PR TITLE
SQL: Extend ScriptTemplate to support functions

### PR DESCRIPTION
**Pull request by costin**
Add support for Painless function definitions to ScriptTemplate for
better script (including chaining).

### CODE REVIEW

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/script/ScriptFunction.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.sql.expression.function.scalar.script;
+
+import java.util.Locale;
+import java.util.Objects;
+
+import static java.lang.String.format;
+
+public class ScriptFunction {
```
Comment (nik9000) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/script/ScriptFunction.java]: Could you add javadoc for this?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/script/ScriptTemplate.java [hunk 2]
```diff
@@ -26,15 +30,25 @@
...
 
     public ScriptTemplate(String template) {
         this(template, Params.EMPTY, DataType.KEYWORD);
     }
 
     public ScriptTemplate(String template, Params params, DataType outputType) {
+        this(template, params, emptyList(), outputType);
+    }
+
+    public ScriptTemplate(String template, Params params, ScriptFunction function, DataType outputType) {
```
Comment (nik9000) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/script/ScriptTemplate.java]: I'd prefer not to have so many ctors. Could you use `singletonList(function)` instead of this? I'd prefer not to have all the old ctors too but that sounds like a good thing to do in a followup.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/script/ScriptTemplate.java [hunk 3]
```diff
@@ -63,12 +81,27 @@ public Script toPainless() {
 
     private String bindTemplate() {
         List<String> binding = params.asCodeNames();
-        return binding.isEmpty() ? template : format(Locale.ROOT, template, binding.toArray());
+
+        String header = null;
+        // sort functions alphabetically
```
Comment (nik9000) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/scalar/script/ScriptTemplate.java]: I *think* I know why you want them sorted alphabetically but I think it'd be nice to have it in a comment.

File: x-pack/qa/sql/src/main/java/org/elasticsearch/xpack/qa/sql/jdbc/CsvSpecTestCase.java [hunk 4]
```diff
@@ -6,6 +6,7 @@
 package org.elasticsearch.xpack.qa.sql.jdbc;
 
 import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
```
Comment (nik9000) [inline x-pack/qa/sql/src/main/java/org/elasticsearch/xpack/qa/sql/jdbc/CsvSpecTestCase.java]: Can you drop this unrelated change?

**Review by nik9000 — CHANGES_REQUESTED** [review]

**Comment by costin** [comment]
Replaced by #30681
